projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,256 @@
+
+#include <unistd.h>
+#include <sys/stat.h>
+#include <ctime>
+
+
+
+
+
+
+// Standard C++ includes
+#include <iostream>
+#include <sstream>
+
+using namespace BoBRobotics;
+using namespace BoBRobotics::ImgProc;
+using namespace BoBRobotics::Video;
+using namespace std::literals;
+
+
+void readCameraThreadFunc(BoBRobotics::Video::Input *cam,
+                        cv::Mat &img,
+                        std::atomic<bool> &shouldQuit,
+                        std::mutex &mutex)
+{
+
+    while(!shouldQuit) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        std::lock_guard<std::mutex> lock(mutex);
+        cam->readFrame(img);
+    }
+}
+
+int bobMain(int argc, char* argv[])
+{
+    constexpr auto UPDATE_INTERVAL = 100ms;
+    using degree_t = units::angle::degree_t;
+
+    // setting up
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    BoBRobotics::Robots::RCCarBot bot;
+    BoBRobotics::GPS::Gps gps;
+    gps.connect(path_linux);
+    BoBRobotics::BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+    std::atomic<bool> shouldQuit{false};
+    std::mutex mex;
+
+    time_t rawtime;
+    struct tm * timeinfo;
+    char buffer[80];
+    time (&rawtime);
+    timeinfo = localtime(&rawtime);
+    strftime(buffer,sizeof(buffer),"%d-%m-%Y",timeinfo);
+    std::string str(buffer);
+
+    int seconds_to_run = 30; // default value
+    if (argc > 1) {
+        seconds_to_run = stoi(std::string(argv[1]));
+        LOGD << "running for " << seconds_to_run << " seconds";
+    }
+
+    // check to see if we get valid coordinates by checking GPS quality
+    int gps_hour,gps_minute,gps_second;
+    BoBRobotics::GPS::TimeStamp gps_time;
+    bool found = false;
+    while (numTrials > 0 && !found) {
+        try {
+            if (gps.getGPSData().gpsQuality != BoBRobotics::GPS::GPSQuality::INVALID) {
+
+                std::this_thread::sleep_for(std::chrono::milliseconds(2000));
+                // getting time data to set folder name
+                gps_time =  gps.getGPSData().time;
+                gps_hour = gps_time.hour;
+                gps_minute = gps_time.minute;
+                gps_second = gps_time.second;
+                found = true;
+                LOGD << " we have a valid measurement";
+
+            } else {
+                numTrials--;
+            }
+
+        } catch(BoBRobotics::GPS::GPSError &e) {
+            LOGW << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]";
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+
+    if (numTrials == 0) {
+        LOGW << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program ";
+        return EXIT_FAILURE;
+    }
+    std::stringstream ss;
+    ss << "imgdataset_"  << str << "_" << gps_hour << "-" << gps_minute << "-" << gps_second;
```

Comment:
On Jenkins, there's a warning about these variables possibly being used uninitialised. Do you get a compiler warning too? (The issue is that if there is no GPS signal, the variables never get set and then they're used here anyway.)

FYI: Jenkins does test if there are compiler warnings and it'll mark things as failing if there are. You *do* want to fix them because a) they might be indicative of a real problem or b) they are harmless, but they obscure other warning messages that are real problems.